surfacers/prometheus/prometheus.go

```diff
@@ -379,6 +382,7 @@ func (ps *PromSurfacer) record(em *metrics.EventMetrics) {
 
 // writeData writes metrics data on w io.Writer
 func (ps *PromSurfacer) writeData(w io.Writer) {
+	ps.deleteExpiredMetrics()
```

Comment:
Thank you for the review.

So, is it better to delete the data by default in select?

``` go
select {
  case <-ctx.Done():
    ps.l.Infof("Context canceled, stopping the input/output processing loop.")
    return
  case em := <-ps.emChan:
    ps.record(em)
  case hw := <-ps.queryChan:
    ps.writeData(hw.w)
    close(hw.doneChan)
  default:
    ps.deleteExpiredMetrics()
}
```

I fully understood what you said.

But the first thing I thought was that if I let it work in the default code, more calls would occur. That's why we expect more overhead.

I thought deleting data on call was a better way than checking data even when there was no data to erase.

Reply:
I try to apply code review... But I had some problem about the test code.

If I change it like the code I uploaded above, it will be a little difficult to write the test code.

I took a error message about concurrency issue map. 

``` console
fatal error: concurrent map read and map write
```

The reason is to check and delete expired data from the moment PromSurfacer is created. However, if an event is not generated on a channel and we will call `ps.record()` directly, the test will not work because of concurrency issue like above.

To solve these issues, I try to putting data into a channel, but this logic should act asynchronously, so it's not clear which one will run first. (`case hw := <-ps.queryChan:` vs `  default:` )

Do you have any good ideas for writing a test code for the above issues?

Reply:
Sorry for the much delayed response. I was not suggesting to delete stale data points in select-default, but instead, there should be a different timer (say one that runs every 10 min), which should trigger the deletion of stale timeseries. Does it make sense?